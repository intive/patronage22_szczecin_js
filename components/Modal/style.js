import styled from 'styled-components'
import Icon from '../Icon/Icon'
import Button from '../Button/Button'

export const ModalOverlay = styled.div`
  display: none;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  background: #515151;
  opacity: 0.32;
  z-index: 20;

  @media (min-width: 768px) {
    display: block;
  }
`

export const ModalContainer = styled.div`
  display: flex;
  flex-direction: column;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: -webkit-fill-available;
  padding: 2rem;
  background: #FFFFFF;
  z-index: 20;

  @media (min-width: 768px) {
    top: 50%;
    left: 50%;
    transform: translate(-50%,-50%);
    padding: 3.5rem 3rem;
    width: auto;
    height: auto;
    max-height: 60vh;
    box-shadow: 0px 8px 16px rgba(98, 98, 98, 0.24);
    border-radius: 8px;
  }
`
export const ModalHeader = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`

export const StyledIcon = styled(Icon)`
  display: none;
  justify-content: center;
  align-items: center;
  margin-bottom: 2rem;
  width: 112px;
  height: 112px;
  margin-top: -7.5rem;
  border-radius: 50%;
  background: #E1E1E1;
  color: #515151;
  font-size: 4rem;

  @media (min-width: 768px) {
    display: flex;
  }
`
export const ModalTitle = styled.p`
  margin-bottom: 0.5rem;
  font-size: 1.5rem;
  line-height: 2rem;
  color: #515151;
`

export const ModalSubtitle = styled.p`
  text-align: center;
  font-size: 1rem;
  line-height: 1.5rem;
  color: #9E9E9E;
`

export const ModalContent = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  flex-grow: 1;
  overflow: auto;
`

export const ModalFooter = styled.div`
  display: flex;
  justify-content: space-around;
  margin-top: 1rem;
  height: 2.5rem;
  width: 100%;


  @media (min-width: 768px) {
    position: static;
    justify-content: flex-end;
  }
`

export const StyledCancelButton = styled(Button)`
  display: flex;
  justify-content: center;
  width: 45%;

  @media (min-width: 768px) {
    width: auto;
  }
`

export const StyledContinueButton = styled(Button)`
  display: flex;
  justify-content: center;
  width: 45%;

  @media (min-width: 768px) {
    width: auto;
  }
`
