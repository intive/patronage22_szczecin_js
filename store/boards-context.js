import { createContext, useState } from 'react'
import { getBoards } from '../services/internal-api'

const BoardsContext = createContext({
  boards: [],
  reload: () => {}
})

export const BoardsContextProvider = (props) => {
  const [updateBoards, setUpdateBoards] = useState(props.boards)

  const reloadHandler = async () => {
    setUpdateBoards((await getBoards()).data)
  }

  const context = {
    boards: updateBoards,
    reload: reloadHandler
  }

  return (
    <BoardsContext.Provider value={context}>
      {props.children}
    </BoardsContext.Provider>
  )
}

export default BoardsContext
