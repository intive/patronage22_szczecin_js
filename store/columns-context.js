import { createContext, useState } from 'react'
import { getBoard } from '../services/internal-api'

const ColumnsContext = createContext({
  columns: [],
  reload: () => {}
})

export const ColumnsContextProvider = (props) => {
  const [updateColumns, setUpdateColumns] = useState(props.columns)

  const reloadHandler = async () => {
    setUpdateColumns((await getBoard(props.id)).data.columns)
  }

  const context = {
    columns: updateColumns,
    reload: reloadHandler
  }

  return (
    <ColumnsContext.Provider value={context}>
      {props.children}
    </ColumnsContext.Provider>
  )
}

export default ColumnsContext
