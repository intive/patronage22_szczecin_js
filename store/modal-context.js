import { createContext } from 'react'

const ModalContext = createContext({})

export const ModalProvider = ModalContext.Provider

export default ModalContext
