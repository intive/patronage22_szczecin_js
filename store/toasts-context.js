import { createContext } from 'react'
import useToast from '../hooks/useToast'

const ToastsContext = createContext({
  showSuccessToast: () => {},
  showErrorToast: () => {}
})

export const ToastsContextProvider = (props) => {
  const { showToast, RenderToast, isOpenToast } = useToast()

  const showSuccessToastHandler = async (notification) => {
    showToast('success', notification)
  }

  const showErrorToastHandler = async (notification) => {
    showToast('error', notification)
  }

  const context = {
    showSuccessToast: showSuccessToastHandler,
    showErrorToast: showErrorToastHandler
  }

  return (
    <ToastsContext.Provider value={context}>
      {props.children}
      {isOpenToast && <RenderToast />}
    </ToastsContext.Provider>
  )
}

export default ToastsContext
