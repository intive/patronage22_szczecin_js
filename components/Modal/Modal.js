import { ModalOverlay, ModalContainer, ModalHeader, StyledIcon, ModalTitle, ModalSubtitle, ModalContent, ModalFooter, StyledCancelButton, StyledContinueButton } from './style'
import { useTranslation } from 'next-i18next'
import { useEffect } from 'react'
import ReactPortal from '../ReactPortal/ReactPortal'

export default function Modal ({ children, isOpen, handleClose, icon, title, subtitle }) {
  const { t } = useTranslation(['common'])

  useEffect(() => {
    const closeOnEscapeKey = e => e.key === 'Escape' ? handleClose() : null

    if (isOpen) {
      document.body.style.overflow = 'hidden'
      window.addEventListener('keydown', closeOnEscapeKey)
    }

    return () => {
      document.body.style.overflow = ''
      window.removeEventListener('keydown', closeOnEscapeKey)
    }
  }, [isOpen, handleClose])

  if (!isOpen) return null

  return (
    <ReactPortal wrapperId='modal-portal'>
      <ModalOverlay data-testid='backdrop' onClick={handleClose} />
      <ModalContainer aria-hidden='false' role='dialog' data-testid='modal'>
        <ModalHeader>
          {icon && <StyledIcon name={icon} />}
          <ModalTitle id='modalTitle'>{title}</ModalTitle>
          <ModalSubtitle id='modalSubtitle'>{subtitle}</ModalSubtitle>
        </ModalHeader>
        <ModalContent>{children}</ModalContent>
        <ModalFooter>
          <StyledCancelButton text onClick={handleClose}>{t('buttons.cancel')}</StyledCancelButton>
          <StyledContinueButton onClick={handleClose}>{t('buttons.continue')}</StyledContinueButton>
        </ModalFooter>
      </ModalContainer>
    </ReactPortal>
  )
}
