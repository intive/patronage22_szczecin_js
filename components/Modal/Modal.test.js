/* globals describe, it, expect  */
/* eslint-env jest */

import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ModalProvider } from '../../store/modal-context'
import Modal from './Modal'

const renderModal = (value) => render(
  <ModalProvider value={value}>
    <Modal />
  </ModalProvider>
)

describe('Modal', () => {
  it('should render Modal correctly', () => {
    const value = {
      isModalOpen: 'true',
      title: 'title',
      subtitle: 'subtitle'
    }
    const { baseElement } = renderModal(value)

    expect(baseElement).toMatchSnapshot()
  })

  it('should close Modal after click on cancel button', () => {
    const mockOnClick = jest.fn()
    const value = {
      isModalOpen: 'true',
      handleClose: mockOnClick
    }
    renderModal(value)
    userEvent.click(screen.getByRole('button', { name: /buttons.cancel/i }))

    expect(mockOnClick).toHaveBeenCalledTimes(1)
  })

  it('if button disabled, should not close Modal after click on save button', () => {
    const mockOnClick = jest.fn()
    const value = {
      isModalOpen: 'true',
      closeModal: mockOnClick,
      disabled: 'disabled'
    }
    renderModal(value)
    expect(screen.queryByRole('button', { name: /buttons.save/i })).toBeDisabled()
    userEvent.click(screen.getByRole('button', { name: /buttons.save/i }))

    expect(mockOnClick).toHaveBeenCalledTimes(0)
  })

  it('should close Modal after click on Backdrop', () => {
    const mockOnClick = jest.fn()
    const value = {
      isModalOpen: 'true',
      handleClose: mockOnClick
    }
    renderModal(value)
    userEvent.click(screen.getByTestId('backdrop'))

    expect(mockOnClick).toHaveBeenCalledTimes(1)
  })

  it('should close Modal after click on Esc keyboard button', () => {
    const mockOnClick = jest.fn()
    const value = {
      isModalOpen: 'true',
      handleClose: mockOnClick
    }
    renderModal(value)
    userEvent.keyboard('{esc}')

    expect(mockOnClick).toHaveBeenCalledTimes(1)
  })
})
