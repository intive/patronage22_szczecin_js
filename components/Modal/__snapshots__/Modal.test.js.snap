// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render Modal correctly 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    id="modal-portal"
  >
    <div
      class="sc-dkPtRN hBZupT"
      data-testid="backdrop"
    />
    <div
      aria-hidden="false"
      class="sc-hKwDye loeQtZ"
      role="dialog"
    >
      <div
        class="sc-eCImPb jdRPcx"
      >
        <p
          class="sc-gKclnd eVXqrc"
          id="modalTitle"
        >
          title
        </p>
        <p
          class="sc-iCfMLu hKohOx"
          id="modalSubtitle"
        >
          subtitle
        </p>
      </div>
      <div
        class="sc-furwcr flDNAI"
      />
      <div
        class="sc-pVTFL hIvRrS"
      >
        <button
          class="sc-bdvvtL jVCSJw sc-jrQzAO bsfgRs"
        >
          buttons.cancel
        </button>
        <button
          class="sc-bdvvtL dRBcfm sc-kDTinF gcMRAr"
        >
          buttons.continue
        </button>
      </div>
    </div>
  </div>
</body>
`;
